### PR TITLE
Remove unused internal functions

### PR DESCRIPTION
**Pull request by akiss77**
The internals have surprisingly large number of unused functions at
surprising places. This patch cleans them out to make maintenance
easier and to prevent dead code having effect on future decisions.

The patch intentionally does not try to clean up public API but
focuses on internal modules only that are/should not be reachable
from "outside".

However, unit tests do access private API, thus tests of literal
storage had to be adjusted.

JerryScript-DCO-1.0-Signed-off-by: Akos Kiss akiss@inf.u-szeged.hu

### CODE REVIEW

**Comment by akiss77** [comment]
I'm happy to discuss whether any/some of these should be kept in the code base. But the fact that these functions are there and not used at all certainly signal some issue.

**Comment by LaszloLango** [comment]
Is it possible to compare the built binary with and without LTO and search for eliminated functions in DCE phase?

**Comment by akiss77** [comment]
If we used gcc to find the unused functions we'd get all those functions which are not needed for the actual application, which can be more than what's unnecessary for the library. That is, we would get a library specialized for the application. We don't want that.

Actually, I've used another tool `cflow` to find the functions removed by this patch. I've run
```bash
cflow -d 2 -r `find . -name '*.c'`
```
in jerry-core and then did some manual investigation on the output.

**Comment by akiss77** [comment]
What really strikes me is that `ecma_op_function_declaration` and `ecma_op_put_value_object_base` are completely unused. Can anyone recall when and how these got replaced (and by what)?

**Comment by LaszloLango** [comment]
I think `ecma_op_function_declaration` got unused when CBC branch was merged. `ecma_op_put_value_object_base` may be the same. @zherczeg, any idea? Don't we really need them?

File: jerry-core/ecma/base/ecma-helpers-string.c [hunk 1]
```diff
@@ -1351,31 +1318,6 @@ ecma_is_string_magic_longpath (const ecma_string_t *string_p, /**< ecma-string *
...
- *
- * Converts passed ecma-string to zt-string and
- * checks if it is equal to one of magic string
- *
- * @return true - if magic string equal to passed string was found,
- *         false - otherwise.
- */
-static bool
-ecma_is_ex_string_magic_longpath (const ecma_string_t *string_p, /**< ecma-string */
-                                  lit_magic_string_ex_id_t *out_id_p) /**< [out] external magic string's id */
```
Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-string.c]: External strings are needed, so I would not remove these two functions.

File: jerry-core/ecma/operations/ecma-lex-env.c [hunk 2]
```diff
@@ -82,21 +82,6 @@ ecma_get_global_environment (void)
...
- *         false - otherwise.
- */
-bool
-ecma_is_lexical_environment_global (ecma_object_t *lex_env_p) /**< lexical environment */
-{
-  JERRY_ASSERT (lex_env_p != NULL
-                && ecma_is_lexical_environment (lex_env_p));
-
-  return (lex_env_p == ecma_global_lex_env_p);
-} /* ecma_is_lexical_environment_global */
```
Comment (zherczeg) [inline jerry-core/ecma/operations/ecma-lex-env.c]: Perhaps this one has an API dependency?

**Comment by zherczeg** [comment]
Yes, CBC made them unused. Do we gain a binary size?

File: jerry-core/ecma/base/ecma-helpers-string.c (hunk 1 above)
Comment (akiss77) [inline jerry-core/ecma/base/ecma-helpers-string.c]: Yeah, that's a tricky one... The catch is that these two functions have almost no effect on external magic string support. If you check their "non-external" counterpart, `ecma_is_string_magic`, you can find out that it's used at a single place only: in `TRY_TO_INSTANTIATE_PROPERTY_ROUTINE_NAME` (in ecma-builtin-internal-routines-template.inc.h). There we try to find a magic string id associated with a builtin. All that is basically hard-wired/built into jerry. And there is no other use for `ecma_is_string_magic`. And I cannot see any valid scenario where external magic strings could be used in a similar fashion.

The real handling of magic strings is in `lit/`. The truly useful functions, like `lit_is_utf8_string_magic` or `it_is_ex_utf8_string_magic` is not touched by this patch.

File: jerry-core/ecma/operations/ecma-lex-env.c (hunk 2 above)
Comment (akiss77) [inline jerry-core/ecma/operations/ecma-lex-env.c]: This was used by `ecma_op_function_declaration` only. There is no API equivalent for this.

**Comment by akiss77** [comment]
@zherczeg: With LTO=ON, size gain is only visible in debug builds (cca. 6-8K text), but with LTO=OFF, ~1400 bytes are saved on arm-linux and arm-mcu even in release (also, 2.8K saved on x64-linux).

Details (LTO=OFF):
```
binary                                | data (mas) | data (rem) | diff | diff[%] | text (mas) | text (rem) | diff | diff[%]
--------------------------------------|------------|------------|------|---------|------------|------------|------|--------
debug.linux-arm7lhf                   |         28 |         28 |      |         |     623276 |     616748 | 6528 |   1.05%
debug.linux-arm7lhf-cp                |         24 |         24 |      |         |     451956 |     445684 | 6272 |   1.39%
debug.linux-arm7lhf-cp_minimal        |         24 |         24 |      |         |     306936 |     300668 | 6268 |   2.04%
debug.linux-arm7lhf-mem_stats         |         28 |         28 |      |         |     625708 |     619180 | 6528 |   1.04%
debug.linux-arm7lhf-mem_stress_test   |         28 |         28 |      |         |     623292 |     616764 | 6528 |   1.05%
debug.linux-x86_64                    |         48 |         48 |      |         |     781227 |     772481 | 8746 |   1.12%
debug.linux-x86_64-cp                 |         32 |         32 |      |         |     596002 |     587660 | 8342 |   1.40%
debug.linux-x86_64-cp_minimal         |         32 |         32 |      |         |     403145 |     394803 | 8342 |   2.07%
debug.linux-x86_64-mem_stats          |         48 |         48 |      |         |     784242 |     775496 | 8746 |   1.12%
debug.linux-x86_64-mem_stress_test    |         48 |         48 |      |         |     781247 |     772501 | 8746 |   1.12%
debug.mcu_stm32f4                     |         48 |         48 |      |         |     633728 |     627160 | 6568 |   1.04%
debug.mcu_stm32f4-cp                  |         44 |         44 |      |         |     457208 |     450904 | 6304 |   1.38%
debug.mcu_stm32f4-cp_minimal          |         44 |         44 |      |         |     303480 |     297180 | 6300 |   2.08%
release.linux-arm7lhf                 |         28 |         28 |      |         |     197916 |     196476 | 1440 |   0.73%
release.linux-arm7lhf-cp              |         24 |         24 |      |         |     150804 |     149468 | 1336 |   0.89%
release.linux-arm7lhf-cp_minimal      |         24 |         24 |      |         |      93498 |      92162 | 1336 |   1.43%
release.linux-arm7lhf-mem_stats       |         28 |         28 |      |         |     198852 |     197412 | 1440 |   0.72%
release.linux-arm7lhf-mem_stress_test |         28 |         28 |      |         |     197932 |     196492 | 1440 |   0.73%
release.linux-x86_64                  |         48 |         48 |      |         |     315877 |     313012 | 2865 |   0.91%
release.linux-x86_64-cp               |         32 |         32 |      |         |     250368 |     247709 | 2659 |   1.06%
release.linux-x86_64-cp_minimal       |         32 |         32 |      |         |     160806 |     158147 | 2659 |   1.65%
release.linux-x86_64-mem_stats        |         48 |         48 |      |         |     317155 |     314290 | 2865 |   0.90%
release.linux-x86_64-mem_stress_test  |         48 |         48 |      |         |     315908 |     313043 | 2865 |   0.91%
release.mcu_stm32f3                   |         48 |         48 |      |         |     206784 |     205304 | 1480 |   0.72%
release.mcu_stm32f3-cp                |         44 |         44 |      |         |     155456 |     154112 | 1344 |   0.86%
release.mcu_stm32f3-cp_minimal        |         44 |         44 |      |         |      91216 |      89872 | 1344 |   1.47%
release.mcu_stm32f4                   |         48 |         48 |      |         |     206784 |     205304 | 1480 |   0.72%
release.mcu_stm32f4-cp                |         44 |         44 |      |         |     155456 |     154112 | 1344 |   0.86%
release.mcu_stm32f4-cp_minimal        |         44 |         44 |      |         |      91216 |      89872 | 1344 |   1.47%
```

File: jerry-core/ecma/base/ecma-helpers-string.c (hunk 1 above)
Comment (LaszloLango) [inline jerry-core/ecma/base/ecma-helpers-string.c]: I think `ecma_is_ex_string_magic` and `ecma_is_ex_string_magic_longpath` are unnecessary and never really used. I agree with removing them.

Comment (zherczeg) [inline jerry-core/ecma/base/ecma-helpers-string.c]: If it does really not hurt external string support, it is ok to remove them.

**Comment by zherczeg** [comment]
LGTM

**Comment by LaszloLango** [comment]
LGTM

**Pull request merged**